File: Algorithm/problem/21_Hash/베스트앨범.py
# 주요 정보 : 재생 횟수, 장르, 고유번호
def solution(genres, plays):
    gen_play = {}
    gen_song = {}
    answer = []
    
    for index, (genre, play) in enumerate(zip(genres, plays)):
        gen_play[genre] = gen_play.get(genre, 0) + play
        
        if not genre in gen_song.keys():
            gen_song[genre] = []
            
        gen_song[genre].append((index, play))
        
    for genre, plays in sorted(gen_play.items(), key = lambda x : x[1], reverse=True):
        for index, play in sorted(gen_song[genre], key = lambda x : (-x[1], x[0]))[:2]:
            answer.append(index)
            
    return answer

File: Algorithm/problem/21_Hash/베스트앨범.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    
    unordered_map<string, int> gen_plays;
    unordered_map<string, vector<pair<int, int>>> gen_songs;
    
    for(int i = 0; i < genres.size(); i++){
        gen_plays[genres[i]] += plays[i];    
        gen_songs[genres[i]].push_back({plays[i], i});    
    }
    
    vector<pair<string, int>> gen_order;
    
    for (auto& [genre, total] : gen_plays){
        gen_order.push_back({genre, total});
    }
    
    sort(gen_order.begin(), gen_order.end(), 
         [](const pair<string, int>& a, const pair<string, int>& b){
             return a.second > b.second;
         });
    
    for(auto& [genre, total] : gen_order){
        vector<pair<int, int>>& songs = gen_songs[genre];
        
        sort(songs.begin(), songs.end(), 
            [](const pair<int, int>& a, const pair<int, int>& b){
                if(a.first == b.first){
                    return a.second < b.second;
                }
                return a.first > b.first;
            });
        
        for(int i = 0; i < songs.size() && i < 2; i++){
            answer.push_back(songs[i].second);
        }
    }
    return answer;
}